### PR TITLE
MAYA123876 adding docs about saving layers

### PR DESCRIPTION
**Pull request by pierrebai-adsk**

### CODE REVIEW

File: lib/mayaUsd/nodes/Layer_Saving_Docs.md [hunk 1]
```diff
@@ -0,0 +1,442 @@
...
+
+*Partial answer*: the session layer gets serialized to a text buffer.
+Maybe that affects its ID? Or maybe freshly-created layers don't have an ID?
+
+
+### What are the responsibilites of the various objects?
+
+We need to write more technical documentation to answer this.
+
+
```
Comment (ppt-adsk) [inline lib/mayaUsd/nodes/Layer_Saving_Docs.md:317]: Remove, keep only in internal Wiki.

File: lib/mayaUsd/nodes/Layer_Saving_Docs.md [hunk 2]
```diff
@@ -0,0 +1,442 @@
...
+When the root layer of a freshly created Stage gets saved, its ID
+changes from no-name to file-name. To be able to find the layer again,
+this new ID is set on the Proxy Shape. This allows the Proxy Shape to
+find it again, but it also marks the Proxy Shape dirty, which mean it
+will be re-computed when accessed again. This ID change is done in the
+Layer Manager when it saves the layer. The Layer manager updates its
+internal layer database, so its findLayer() will be able to find the
+renamed root layer, in particular when the Proxy Shape asks for it. The
+problem was that since the root layer had changed ID, it would not be
+found in the Stage caches.
```
Comment (ppt-adsk) [inline lib/mayaUsd/nodes/Layer_Saving_Docs.md:347]: Remove the last sentence, document the new code.

File: lib/mayaUsd/nodes/Layer_Saving_Docs.md [hunk 3]
```diff
@@ -0,0 +1,442 @@
...
+
+*Fix*: update the Stage caches when the root layer ID changes.
+
+
+### The Session Layer ID Problem
+
+The Layer Manager also does this for the session layer. Except there was
```
Comment (ppt-adsk) [inline lib/mayaUsd/nodes/Layer_Saving_Docs.md:354]: Also does what?

File: lib/mayaUsd/nodes/Layer_Saving_Docs.md [hunk 4]
```diff
@@ -0,0 +1,442 @@
...
+a bug: it would only set the session layer ID attribute on the Proxy
+Shape when saving *_all_* layers inside the Maya scene file. If
+non-session layers were saved to external USD files, then the session
+layer ID would _*not*_ get set on the Proxy Shape. This would lead the
+Proxy Shape to not find the session layer. The reason being that the
+root layer would change ID, thus would not be found in the cache, thus a
+new stage would be created with a new session layer.
+
+*Fix:* set the session layer ID on the Proxy Shape when saving to
+external USD files.
```
Comment (ppt-adsk) [inline lib/mayaUsd/nodes/Layer_Saving_Docs.md:364]: Remove most of this, describe the new code.

File: lib/mayaUsd/nodes/Layer_Saving_Docs.md [hunk 5]
```diff
@@ -0,0 +1,442 @@
...
+- ... so we lose the session layer.
+
+An issue here is that we cannot fully ensure that no code modifies the
+Proxy Shape or that no code accesses the Proxy Shape and triggers a
+re-compute. It is very easy to overlook one code path or have a rare
+notification be sent mid-way through the save. The best approach in the
+short term is to try to make the current behaviour sturdy and in
+particular avoid accessing the Proxy Shape multiple times during the
+saving of layer and the Maya scene.
+
```
Comment (ppt-adsk) [inline lib/mayaUsd/nodes/Layer_Saving_Docs.md:398]: Remove, keep only on in internal Wiki.

File: lib/mayaUsd/nodes/Layer_Saving_Docs.md [hunk 6]
```diff
@@ -0,0 +1,442 @@
...
+- Instead, we retrieve all information we will need before-hand, before
+any other code runs, including the callback.
+- This ensure we don't access the proxy shape in the middle, between the
+callback and the rest of the USD layer saving code.
+- Since we retrieve the information, we might as well give it to the
+callback to avoid doing the same retrieval multiple times.
+- It also reduces the probability that the callback would accidentally
+cause a similar bug where it would modify the proxy shape
+  in one part of its code then access it again, triggering a recompute
+that would lose the session.
```
Comment (ppt-adsk) [inline lib/mayaUsd/nodes/Layer_Saving_Docs.md:413]: Remove historical reference.

File: lib/mayaUsd/nodes/Layer_Saving_Docs.md [hunk 7]
```diff
@@ -0,0 +1,442 @@
...
+the Maya scene was saved, the node would already be gone and would not
+get saved.
+
+*Fix*: we now no longer delete the Layer Manager Maya node when we are
+in the middle of a Maya scene save. The reasoning being that while in
+the middle of a save, code will have created the Layer Manager node and
+nothing that would accidentally call findLayer() should cause the node
+to be deleted, since we need to have it to be able to save it into the
+Maya scene, since the layer manager node is how we save layers in Maya.
+(The layers are attributes on the layer manager node.)
```
Comment (ppt-adsk) [inline lib/mayaUsd/nodes/Layer_Saving_Docs.md:442]: Remove historical references, document state of current code.

**Review by ppt-adsk — CHANGES_REQUESTED** [review]
Massive amount of work, thanks for doing this!  In general, I would remove internal details, and re-work historical descriptions and turn them into descriptions of the current state of the code.

File: lib/mayaUsd/nodes/Layer_Saving_Docs.md [hunk 8]
```diff
@@ -82,170 +84,147 @@ is how a newly-created stage is initialized with a root layer.
...
+to find which ones are still dirty and still need to be saved. It then saves
+these layers.
+
+An important addition is that the load-state of USD payload are pure stage-
+level information. That means they do not get saved in the layer data. In
+Maya USD, we do save load state of all payload for each stage as an attribute
+on the Proxy Shape. This is also done in another Maya pre-saved callback.
+
+
+## Save Architecture Challenge
```
Comment (ppt-adsk) [inline lib/mayaUsd/nodes/Layer_Saving_Docs.md:164]: Would word it as "Challenges" plural, for consistency with the text below.

File: lib/mayaUsd/nodes/Layer_Saving_Docs.md [hunk 9]
```diff
@@ -82,170 +84,147 @@ is how a newly-created stage is initialized with a root layer.
...
-Various questions raised by this investigation. For these questions, we
-believe we have a valid answer.
-
-### What stages caches do we have?
-
-*Answer*: See the Stage section above.
-
-
-### Is USD creating them for us? Are we (maya-usd) creating them?
+### Where are stage caches created is Maya USD?
```
Comment (ppt-adsk) [inline lib/mayaUsd/nodes/Layer_Saving_Docs.md:180]: "is Maya USD" --> "in Maya USD".

File: lib/mayaUsd/nodes/Layer_Saving_Docs.md [hunk 10]
```diff
@@ -82,170 +84,147 @@ is how a newly-created stage is initialized with a root layer.
...
-it and actually want ot find a stage wirth the root layer and existing
-session layer, if any are already loaded.
+*Answer*: The goal is that if we create multiple shared stages with the same
+root layer, they will share the session layer. For example, if the user creates
+a new stage, all that the user specifies is the file-name of the root layer
+(and a few flags not relevant for this discussion). In order to find a
+previously-existing shared stage, only the root layer must be passed so that
+USD will return the existing stage, thus implementing the desired sharing of
+the stage and session layer. We could not possibly pass a session layer,
+because we don't have it and we want ot find a stage with the root layer and
```
Comment (ppt-adsk) [inline lib/mayaUsd/nodes/Layer_Saving_Docs.md:221]: "ot" --> "to"

**Review by ppt-adsk — CHANGES_REQUESTED** [review]
Almost there, only very minor typos and details!

File: lib/mayaUsd/nodes/Layer_Saving_Docs.md [hunk 11]
```diff
@@ -82,170 +84,147 @@ is how a newly-created stage is initialized with a root layer.
...
-
-## Possible Improvements
-
-Many of the problems that happened were due to the facts that:
+Layers are saved just before the Maya scene is saved, in a Maya pre-save
+callback. All layers that needs to be saved to external files are saved,
+then all layers that need to be saved inside Maya file, including session
+layers, are serialized into string attributes in the Layer Manager Maya node.
+
+This layer-saving process is not customizable. The USD plugin can provide
```
[the remark this answers is not in the document]
Comment (pierrebai-adsk) [reply lib/mayaUsd/nodes/Layer_Saving_Docs.md:138]: not was left behind, removed.

**Review by ppt-adsk — APPROVED** [review]
Great work, thanks for doing this!

**Comment by ppt-adsk** [comment]
No pre-flight required, documentation only.

**Pull request merged**